### DISCUSSION TITLE
Send a 404 to an API

### DISCUSSION THREAD

**Discussion by andrewfairlie** [vercel/next.js#19738] [Help] [answered]
Is it possible with Next/Vercel to send 404 to an API instead of a page?

We ask because our CMS allows for us to define redirects, and we would like to check "If this page 404s, check if the URI is in our redirects list and if so redirect, else render the 404 page".

I would have ideally done this in `getStaticProps` on `404.js` and then returned a `redirect` in the returned object, but we're unable to identify the URI of the 404'd page on `404.js` unfortunately (unless I'm missing something?)

**Comment by Tomburgs** [comment, marked as answer]
Have a look at the [documentation for data fetching](https://nextjs.org/docs/basic-features/data-fetching).

On both `getStaticProps` & `getServerSideProps` you can return `notFound` which would return 404, or `redirect` which would redirect the user, then you could create a catch-all route, such as `[...slug].tsx` and see if you have content & redirect for this page.

However, if you're purely doing redirects, you can also have a look at the [redirects documentation](https://nextjs.org/docs/api-reference/next.config.js/redirects).

**Reply by andrewfairlie** [reply]
Hi @Tomburgs this is brilliant, thank you. I knew all of those techniques but didn't connect the dots.

I made `[...handleError].js` on the top level which includes

```
import { returnStaticProps } from "@/lib/return-static-props"

/*
    This template catches all URIs not otherwise handled by dynamic paths, which means they're invalid URLs
    but before we send them to a 404 page we check if that URL is in Retour, and if it is redirect instead.
*/

function handleError() {
    // We never actually render this page, so its an empty fragment
    return <></>
}
export async function getStaticPaths() {
    return {
        paths: [],
        fallback: "blocking",
    }
}
export async function getStaticProps({ params }) {
    return await returnStaticProps({}, true, `/${params.handleError.join("/")}`)
}

export default handleError
```

`returnStaticProps` is a little helper I wrote to generate my return statement (only because makes it a bit easier to do this logic across pages).

```
import { craft, gql } from "@/lib/craft"

export const returnStaticProps = async (props, errored, uri) => {
    let returnStatement = {
        revalidate: 1,
    }

    // We *do* have an entry, let it through
    if (!errored) {
        returnStatement.props = props
        return returnStatement
    }

    // Check Retour to see if a redirect exists for this
    const retour = await craft(
        gql`
            {
                retour(uri: "${uri}") {
                    redirectDestUrl
                    redirectHttpCode
                }
            }
        `,
        null,
        null
    )

    if (retour.retour !== null) {
        const redirect = retour.retour

        returnStatement.redirect = {
            destination: redirect.redirectDestUrl,
            permanent: redirect.redirectHttpCode === 301,
        }

        return returnStatement
    }

    // We can't find anything, 404
    returnStatement.notFound = true
    return returnStatement
}
```


I also use that function on actual pages too. For example on `products/category/[slug].js` I use...

```
    return await returnStaticProps(
        {
            entry: craftCategoryQuery.entry,
            products: craftProductsQuery.products,
            finishes: craftFinishesQuery.finishes,
            globals,
        },
        craftCategoryQuery.entry === null,
        `/products/category/${params.slug}`
    )
```
